File: ui/src/sw-config.ts
/**
 * Pure helpers for the dashboard's PWA service worker.
 *
 * The SW fetches ``/sw-config.json`` from the controller on
 * install/update; the response shape is described by ``SwConfig``
 * below. All decision logic (should the SW serve the SPA shell for
 * THIS navigation? should it pass through to the network?) lives in
 * pure functions here so they can be unit-tested without needing
 * the ServiceWorkerGlobalScope.
 *
 * Why runtime config instead of build-time regex
 * ----------------------------------------------
 * The SW used to ship with hardcoded ``navigateFallbackDenylist``
 * regexes ("``/api/``" and "``/app/(?!media-stack-ui(?:/|$))``").
 * That coupled the bundle to the deployment shape — operators who
 * renamed the dashboard mount via the routing-admin page got the
 * old "SW hijacks sister apps" bug back.
 *
 * Pulling the patterns from the controller's routing engine means
 * the SW always tracks the live deployment. A single source of
 * truth, and a renamed prefix takes effect on the next SW update.
 */

/** Live shape of ``GET /sw-config.json``. */
export interface SwConfig {
  version: number;
  /** Dashboard's mount point, e.g. ``"/app/media-stack-ui"``. No
   *  trailing slash. */
  basepath: string;
  /** Regex strings the SW must NOT serve the SPA shell for. */
  denylist_patterns: string[];
  /** Convenience: every ``/app/<service>`` path the dashboard owns
   *  (today just the basepath; reserved for sub-apps). */
  allowed_app_prefixes: string[];
  /** Every other ``/app/<service>`` deployed alongside the
   *  dashboard. Surfaced for telemetry / future SW logic. */
  sister_app_prefixes: string[];
}

/** Defaults used when the endpoint is unreachable. Match the
 *  pre-runtime-config behavior so first-install offline still
 *  renders the SPA shell. */
export const SW_CONFIG_DEFAULTS: SwConfig = {
  version: 1,
  basepath: "/app/media-stack-ui",
  denylist_patterns: [
    "^/api/",
    "^/app/(?!media-stack-ui(?:/|$))",
  ],
  allowed_app_prefixes: ["/app/media-stack-ui"],
  sister_app_prefixes: [],
};

/** Best-effort fetch; falls back to defaults on any error so
 *  ``install`` never blocks. */
export async function fetchSwConfig(
  fetchImpl: typeof fetch,
): Promise<SwConfig> {
  try {
    const res = await fetchImpl("/sw-config.json", { cache: "no-store" });
    if (!res.ok) return SW_CONFIG_DEFAULTS;
    const raw = (await res.json()) as Partial<SwConfig>;
    return normalizeSwConfig(raw);
  } catch {
    return SW_CONFIG_DEFAULTS;
  }
}

/** Coerce a partially-populated payload into a full SwConfig. Any
 *  field that's missing or malformed gets the corresponding default. */
export function normalizeSwConfig(raw: Partial<SwConfig> | null | undefined): SwConfig {
  if (!raw || typeof raw !== "object") return SW_CONFIG_DEFAULTS;
  const basepath = typeof raw.basepath === "string" && raw.basepath
    ? trimTrailingSlash(raw.basepath)
    : SW_CONFIG_DEFAULTS.basepath;
  const denylist = Array.isArray(raw.denylist_patterns)
    ? raw.denylist_patterns.filter((p): p is string => typeof p === "string")
    : SW_CONFIG_DEFAULTS.denylist_patterns;
  const allowed = Array.isArray(raw.allowed_app_prefixes)
    ? raw.allowed_app_prefixes.filter((p): p is string => typeof p === "string")
    : [basepath];
  const sister = Array.isArray(raw.sister_app_prefixes)
    ? raw.sister_app_prefixes.filter((p): p is string => typeof p === "string")
    : [];
  return {
    version: typeof raw.version === "number" ? raw.version : 1,
    basepath,
    denylist_patterns: denylist,
    allowed_app_prefixes: allowed,
    sister_app_prefixes: sister,
  };
}

function trimTrailingSlash(p: string): string {
  return p.replace(/\/+$/, "");
}

/** Compile the regex strings into an array of ``RegExp``. Bad
 *  patterns are filtered out (they shouldn't ship from a healthy
 *  controller, but we don't want one bad regex to disable the
 *  whole denylist). */
export function compileDenylist(patterns: readonly string[]): RegExp[] {
  const out: RegExp[] = [];
  for (const p of patterns) {
    try {
      out.push(new RegExp(p));
    } catch {
      // Drop malformed; SW continues with the survivors.
    }
  }
  return out;
}

/**
 * Decide whether a navigation request should be served from the
 * SPA shell (``index.html`` cached by the SW) vs passed through to
 * the network.
 *
 * Rules:
 *
 *  1. Non-navigation requests: the SW's other route handlers
 *     handle them (precache, runtime caches). This function only
 *     answers "should I serve the SPA shell?".
 *
 *  2. URL outside the dashboard's basepath: pass through. Sister
 *     apps live at ``/app/<other>/`` and need their own backend.
 *
 *  3. URL matches any denylist regex: pass through. Covers
 *     ``/api/*`` (controller endpoints) and ``/app/<other>/*``
 *     (caught by the routing-engine-built pattern).
 *
 *  4. Otherwise: serve the SPA shell.
 */
export function shouldServeSpaShell(
  url: URL,
  config: SwConfig,
  denylist: readonly RegExp[],
): boolean {
  // Cross-origin → never our problem.
  if (typeof self !== "undefined" && "location" in self) {
    if (url.origin !== (self as { location: Location }).location.origin) {
      return false;
    }
  }
  // Outside our basepath → not our problem either.
  const base = config.basepath;
  if (base) {
    const inScope =
      url.pathname === base || url.pathname.startsWith(base + "/");
    if (!inScope) return false;
  }
  // Inside our scope but matches a denylist pattern → pass through.
  for (const re of denylist) {
    if (re.test(url.pathname)) return false;
  }
  return true;
}

File: src/media_stack/api/services/sw_config.py
"""Service-worker config endpoint.

The dashboard's PWA service worker needs to know two things at
runtime:

  1. **Its own basepath** — where the dashboard SPA is mounted
     (``/app/media-stack-ui/`` by default; operators can rename it
     via the routing-admin page). The SW intercepts navigations
     under this path and serves ``index.html`` for offline routing.

  2. **The other ``/app/<service>/`` mount points** — every
     deployed sister app the SW should NOT hijack. When the operator
     visits ``/app/sonarr/``, the SW must let the network handle it
     so Envoy routes to the Sonarr container.

Hardcoding either of these in ``vite.config.ts`` couples build to
deployment shape — a custom-prefixed deploy needs a rebuild. By
emitting both as a JSON document the SW fetches at install/update,
the routing engine becomes the single source of truth.
"""

from __future__ import annotations

import logging
import os
from typing import Any

logger = logging.getLogger("controller_api")

from media_stack.services.apps.stack.routing_defaults import (
    DASHBOARD_SERVICE_ID,
)


def get_sw_config() -> dict[str, Any]:
    """Build the service-worker configuration payload.

    Shape::

        {
          "version": 1,
          "basepath": "/app/media-stack-ui",
          "denylist_patterns": [
            "^/api/",
            "^/app/(?!media-stack-ui(?:/|$))"
          ],
          "allowed_app_prefixes": ["/app/media-stack-ui"],
          "sister_app_prefixes": ["/app/sonarr", "/app/jellyfin", ...]
        }

    ``denylist_patterns`` is what the SW plugs straight into
    ``navigationRoute`` matching — pre-compiled regex strings so the
    SW doesn't re-derive them from the prefix lists.
    ``allowed_app_prefixes`` and ``sister_app_prefixes`` are also
    surfaced so future SW logic (e.g. background sync per app) can
    use them without re-parsing the patterns.

    Falls back to safe defaults when the registry can't be loaded —
    a misconfigured controller still serves a usable SW config
    rather than an empty payload.
    """
    basepath = _resolve_dashboard_basepath()
    sister_prefixes = _list_sister_app_prefixes(basepath=basepath)
    denylist_patterns = _build_denylist_patterns(
        basepath=basepath,
        sister_prefixes=sister_prefixes,
    )
    return {
        "version": 1,
        "basepath": basepath,
        "denylist_patterns": denylist_patterns,
        "allowed_app_prefixes": [basepath],
        "sister_app_prefixes": sister_prefixes,
    }


def _resolve_dashboard_basepath() -> str:
    """Return the dashboard's mount point, e.g.
    ``"/app/media-stack-ui"`` (no trailing slash).

    Resolution order:

      1. ``DASHBOARD_BASEPATH_OVERRIDE`` env — escape hatch for
         non-default deploys.
      2. Profile's ``routing.app_path_prefix`` + the dashboard
         service id — the canonical source of truth.
      3. Hardcoded default ``/app/media-stack-ui`` — only hit when
         the profile is unreachable AND the env override isn't set.
    """
    override = os.environ.get("DASHBOARD_BASEPATH_OVERRIDE", "").strip()
    if override:
        return _normalize_path(override)

    app_prefix = _read_app_path_prefix_from_profile()
    return _normalize_path(f"{app_prefix}/{DASHBOARD_SERVICE_ID}")


def _read_app_path_prefix_from_profile() -> str:
    """Best-effort read of ``routing.app_path_prefix`` from the
    bootstrap profile YAML. Returns ``"/app"`` on any error so a
    misconfigured profile doesn't break the SW config endpoint."""
    try:
        # Imported lazily to avoid a circular import via
        # ``api.services.config`` during module load.
        from media_stack.api.services.config._routing import (
            get_routing as _get_routing,
        )
        cfg = _get_routing()
        prefix = (cfg or {}).get("app_path_prefix")
        if isinstance(prefix, str) and prefix.strip():
            return prefix.strip()
    except Exception as exc:  # noqa: BLE001
        logger.debug(
            "[sw_config] falling back to default /app prefix: %s", exc,
        )
    return "/app"


def _list_sister_app_prefixes(*, basepath: str) -> list[str]:
    """Enumerate ``/app/<service>`` prefixes for every registered
    service that ISN'T the dashboard. Used by the SW to recognize
    paths it should pass through to the network (Envoy)."""
    try:
        from media_stack.core.service_registry.registry import SERVICES
    except Exception as exc:  # noqa: BLE001
        logger.debug("[sw_config] registry unavailable: %s", exc)
        return []
    app_prefix = _read_app_path_prefix_from_profile()
    out: list[str] = []
    for s in SERVICES:
        sid = getattr(s, "id", "")
        if not sid or sid == DASHBOARD_SERVICE_ID:
            continue
        prefix = _normalize_path(f"{app_prefix}/{sid}")
        if prefix != basepath:
            out.append(prefix)
    out.sort()
    return out


def _build_denylist_patterns(
    *,
    basepath: str,
    sister_prefixes: list[str],
) -> list[str]:
    """Assemble the regex strings the SW navigates against. Two
    patterns by default:

      * ``^/api/`` — controller REST surface; SW should never
        substitute the SPA shell for a JSON 401/404.
      * ``^/app/(?!<dashboard-segment>(?:/|$))`` — sister-app
        passthrough.

    The second pattern is built from the basepath segments rather
    than hard-coded so a renamed dashboard mount automatically
    reshapes the regex.
    """
    patterns: list[str] = [r"^/api/"]
    # The basepath is something like ``/app/media-stack-ui``; we
    # need the segment that follows ``/app/``.
    parts = basepath.strip("/").split("/")
    if len(parts) >= 2 and parts[0] == "app":
        dashboard_segment = re.escape(parts[1])
        patterns.append(
            rf"^/app/(?!{dashboard_segment}(?:/|$))",
        )
    elif len(parts) == 1:
        # Non-prefixed deploy (everything at root). SW only needs
        # the /api/ exclusion in this case.
        pass
    return patterns


def _normalize_path(p: str) -> str:
    """Trim trailing slashes; collapse repeated slashes; ensure
    leading slash."""
    if not p:
        return ""
    cleaned = "/" + "/".join(seg for seg in p.split("/") if seg)
    return cleaned


# Lazily imported here so the regex helpers above can use it without
# pulling re into the module-load cost.
import re  # noqa: E402
